web/interceptor/access_control.go

```diff
@@ -0,0 +1,181 @@
+	user
+	// group role implies that the user is a course student + a member of the given group.
+	group
+	// student role implies that the user is enrolled in the course with any role.
+	student
+	// teacher: user enrolled in the course with teacher status.
+	teacher
+	// courseAdmin: an admin user who is also enrolled into the course.
+	courseAdmin
+	// admin is the user with admin privileges.
+	admin
+)
+
+// If there are several roles that can call a method, a role with the least privilege must come first.
+var access = map[string]roles{
+	"GetUser":                 {none},
+	"GetCourse":               {none},
+	"GetCourses":              {none},
+	"CreateEnrollment":        {user},
+	"UpdateCourseVisibility":  {user},
+	"GetCoursesByUser":        {user},
+	"UpdateUser":              {user, admin},
+	"GetEnrollmentsByUser":    {user, admin},
+	"GetSubmissions":          {student, group, teacher, courseAdmin},
+	"GetGroupByUserAndCourse": {group, teacher},
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetAssignments":          {student, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"GetRepositories":         {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+// AccessControl checks user information stored in the JWT claims agains the list of roles required to call the method.
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, request interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		req, ok := request.(requestID)
+		// The GetUserByCourse method sends a CourseUserRequest which has no IDs and needs a database query.
+		if !ok {
+			logger.Errorf("%s failed: message type '%s' does not implement IDFor interface",
+				method, reflect.TypeOf(request).String())
+			return nil, ErrAccessDenied
+		}
+		roles, ok := access[method]
```

Comment:
Can it actually happen that `ok == false`? Given that we do a pre-check that all methods are in the `access` map. Moreover, the returned `roles` from a false lookup should give an empty slice, which would not enter the `for range roles` loop below, ending at 178 and returning the same `ErrAccessDenied`.

Reply:
That is, I suggest to remove the `!ok` check and simply do:
```go
roles := access[method]
for _, role := range roles {
...
```
You could even do:
```go
for _, role := range access[method] {
...
```
But maybe that is less clear... Unless the `access` map is renamed to `accessRolesFor[method]` or something.

Reply:
So something like `for _, role := range access[method]` is pretty safe?

Reply:
Done.